test/test_library/unit/test_init.py

```diff
@@ -0,0 +1,1858 @@
+
+        # prepare state
+        lib = Library(self.cache, self.sqlite_db)
+        lib._dependencies = {'dataset1': 'missed-vid'}
+
+        # testing
+        with self.assertRaises(NotFoundError):
+            lib.check_dependencies(download=False)
+
+    def test_gets_bundle_and_raises_NotFoundError_on_faile(self):
+        # first assert signatures of the functions we are going to mock did not change.
+        assert_spec(Library.get, ['self', 'ref', 'remote', 'force', 'cb', 'location'])
+
+        # prepare state
+        lib = Library(self.cache, self.sqlite_db)
+        lib._dependencies = {'dataset1': 'vid1'}
+        fake_get = fudge.Fake()\
+            .expects_call()\
+            .with_args('vid1', cb=arg.any())\
+            .returns(None)
+
+        # testing
+        with fudge.patched_context(lib, 'get', fake_get):
+            with self.assertRaises(NotFoundError):
+                lib.check_dependencies(download=True)
+
+    # .source property tests
+    def test_returns_source_tree(self):
+        # first assert signatures of the functions we are going to mock did not change.
+        assert_spec(
+            SourceTree.__init__,
+            ['self', 'base_dir', 'library', 'logger'])
+
+        # prepare state
+        lib = Library(self.cache, self.sqlite_db)
+        lib.source_dir = 'the-dir'
+
+        fake_init = fudge.Fake('init')\
+            .expects_call()\
+            .with_args(lib.source_dir, lib, lib.logger)
+
+        # testing
+        with fudge.patched_context(SourceTree, '__init__', fake_init):
+            source = lib.source
+            self.assertIsInstance(source, SourceTree)
+
+    # .new_files property tests
+    def test_generates_new_installed_files(self):
+
+        # prepare state
+        lib = Library(self.cache, self.sqlite_db)
+        # add two installed files with new state.
+        FileFactory(type_=Files.TYPE.BUNDLE, state='new')
+        FileFactory(type_=Files.TYPE.PARTITION, state='new')
+
+        # add not installed
+        FileFactory(type_=Files.TYPE.STORE, state='new')
+
+        # add with another state
+        FileFactory(type_=Files.TYPE.BUNDLE)
+
+        # testing
+        new_files_gen = lib.new_files
+
+        # is generator
+        self.assertTrue(hasattr(new_files_gen, 'next'))
+        new_files = [x for x in new_files_gen]
+        self.assertEquals(len(new_files), 2)
+        self.assertEquals(new_files[0].state, 'new')
+        self.assertEquals(new_files[1].state, 'new')
+
+    # .push tests
+    def test_raises_exception_if_upstream_is_empty(self):
+
+        # prepare state
+        lib = Library(self.cache, self.sqlite_db)
+
+        # testing
+        raised = False
+        try:
+            lib.push(None)
+        except Exception as exc:
+            raised = True
+            self.assertIn('without defining a upstream.', exc.message)
+        self.assertTrue(raised)
+
+    # .sync_library tests
+    @unittest.skip(
```

Comment:
That's surprising, hence the assert. Check that one of them is defined to be the other. I forget which one is primary, but it is probably Dataset.LOCATION.